ipq806x: add support for Cisco Meraki MR42/MR52

clayface opened the pull request:
The MR42 and MR52 are two similar IPQ806x based devices from the Cisco Meraki
"Cryptid" series.

Main features:
  MR42
  -  IPQ8068 1.4GHz
  -  512MB RAM
  -  128MB NAND
  -  2x QCA9992 (2.4 & 5GHz)
  -  1x QCA9889 (2.4 & 5GHz)
  -  1x AR8033 PHY
  -  PoE/AC power

  MR52
  -  IPQ8068 1.4GHz
  -  512MB RAM
  -  128MB NAND
  -  2x QCA9994 (2.4 & 5GHz)
  -  1x QCA9889 (2.4 & 5GHz)
  -  2x AR8033 PHYs
  -  PoE/AC power

Install instructions are in the "ipq806x: add support for Cisco Meraki MR42/MR52" commit.

Extra info:
1. The MR42's AR8033 PHY requires multiple GPIO pinctrl configs on pins 6/7, which I wasn't able to replicate in the device tree in spite of using init and default pinctrl states. When this isn't done correctly the PHY will be unusable apart from in 1000Mb/s, EEE off mode. The most convenient way to rectify this is to replace the bootloader with one that runs with the init config for a longer period of time (the GPL sources have disabled code that does this), thus when the kernel's pinctrl config is loaded it will now be in the correct mode allowing 10/100 mb/s to work. EEE must remain off (as in the Meraki OS) so broken-eee properties are added. Changing the bootloader also allows for tftp booting to be enabled.

This issue doesn't affect the MR52 so this step is not necessary.

2. GMAC_AHB_RESET needs to be deasserted in the kernel, though the modified bootloader for the MR42 is also doing this. We need it for the MR52 though.

3. KERNEL_LOADADDR is set to 0x44208000. I wasn't able to boot at 0x42208000.

Signed-off-by: Matthew Hagan <mnhagan88@gmail.com>

clayface committed:
ipq806x: add support for Cisco Meraki MR42/MR52

The MR42 and MR52 are two similar IPQ806x based devices from the Cisco
Meraki "Cryptid" series.

  MR42 main features:
  -  IPQ8068 1.4GHz
  -  512MB RAM
  -  128MB NAND
  -  2x QCA9992 (2.4 & 5GHz)
  -  1x QCA9889 (2.4 & 5GHz)
  -  1x AR8033 PHY
  -  PoE/AC power

  MR52 main features:
  -  IPQ8068 1.4GHz
  -  512MB RAM
  -  128MB NAND
  -  2x QCA9994 (2.4 & 5GHz)
  -  1x QCA9889 (2.4 & 5GHz)
  -  2x AR8033 PHYs
  -  PoE/AC power

(MR42 Only) Installation via diagnostic mode:

If you can successfully complete step 1 then you can continue to install
via this method without having to open the device. Otherwise please use
the standard UART method. Please note that when booting via TFTP, some
Ethernet devices, in particular those on laptops, will not connect in
time, resulting in TFTP boot not succeeding. In this instance it is
advised to connect via a switch.

  1. Hold down reset at power on and keep holding, after around 10 seconds
     if the orange LED changes behaviour to begin flashing, proceed to
     release reset, then press reset two times. Ensure that the LED has
     turned blue. Note that flashing will occur on some devices, but it
     will not be possible to change the LED colour using the reset button.
     In this case it will still be possible to continue with this install
     method.

  2. Set your IP to 192.168.1.250. Set up a TFTP server serving
     mr42_u-boot.mbn and
     openwrt-ipq806x-generic-meraki_mr42-initramfs-fit-uImage.itb, obtained
     from [1].

  3. Use telnet and connect to 192.168.1.1. Run the following commands to
     install u-boot. Note that all these commands are critical, an error
     will likely render the device unusable.

     Option 3.1:
       If you are sure you have set up the TFTP server correctly you can
       run this script on the device. This will download and flash the
       u-boot image immediately:

       `/etc/update_uboot.sh 192.168.1.250 mr42_u-boot.mbn`

       Once completed successfully, power off the device.

     Option 3.2:
       If you are unsure the TFTP server is correctly set up you can
       obtain the image and flash manually:

       3.2.1. `cd /tmp`
       3.2.2. `tftp-hpa 192.168.1.250 -m binary -c get mr42_u-boot.mbn`
       3.2.3. Confirm file has downloaded correctly by comparing the
              md5sum:

            `md5sum mr42_u-boot.mbn`

       3.2.4. The following are the required commands to write the image.

            `echo 1 > /sys/devices/platform/msm_nand/boot_layout
             mtd erase /dev/mtd1
             nandwrite -pam /dev/mtd1 mr42_u-boot.mbn
             echo 0 > /sys/devices/platform/msm_nand/boot_layout`

          Important: You must observe the output of the `nandwrite`
          command. Look for the following to verify writing is occurring:

            `Writing data to block 0 at offset 0x0
             Writing data to block 1 at offset 0x20000
             Writing data to block 2 at offset 0x40000`

          If you do not see this then do not power off the device. Check
          your previous commands and that mr42_u-boot.mbn was downloaded
          correctly. Once you are sure the image has been written you
          can proceed to power off the device.

  4. Hold the reset button and power on the device. This will immediately
     begin downloading the appropriate initramfs image and boot into it.

     Note: If the device does not download the initramfs, this is likely
     due to the interface not being brought up in time. Changing Ethernet
     source to a router or switch will likely resolve this. You can also
     try manually setting the link speed to 10Mb/s Half-Duplex.

  5. Once a solid white LED is displayed on the device, continue to the
     UART installation method, step 6.

Standard installation via UART - MR42 & MR52

  1. Disassemble the device and connect a UART header. The header pinout
     is as follows:

       1 - 3.3v
       2 - TXD
       3 - RXD
       4 - GND

     Important: You should only connect TXD, RXD and GND. Connecting
     3.3v may damage the device.

  2. Set your IP to 192.168.1.250. Set up a TFTP server serving
     openwrt-ipq806x-generic-meraki_(mr42|mr52)-initramfs-fit-uImage.itb.
     Separately obtain the respective sysupgrade image.

  3. Run the following commands, preferably from a Linux host. The
     mentioned files, including ubootwrite.py and u-boot images, can be
     obtained from [1].

       `python ubootwrite.py --write=(mr42|mr52)_u-boot.bin`

     The default for "--serial" option is /dev/ttyUSB0.

  4. Power on the device. The ubootwrite script will upload the image to
     the device and launch it. The second stage u-boot will in turn load
     the initramfs image by TFTP, provided the TFTP server is running
     correctly. This process will take about 13 minutes. Once a solid
     white LED is displayed, the image has successfully finished
     loading. Note: If the image does not load via TFTP, try again with
     the Ethernet link to 10Mb/s Half-Duplex.

  5. (MR42 only) Do not connect over the network. Instead connect over
     the UART using minicom or similar tool. To replace u-boot with
     the network enabled version, please run the following commands.
     Note that in the provided initramfs images, the u-boot.mbn file
     is located in /root:

     If you have not used the provided initramfs, you must ensure you
     are using an image with "boot_layout" ECC configuration enabled in
     the Kernel. This will be version 5.10 or higher. If you do not do
     this correctly the device will be bricked.

       `insmod mtd-rw i_want_a_brick=1
        mtd erase /dev/mtd8
        nandwrite -pam /dev/mtd8 /root/mr42_u-boot.mbn`

     After running nandwrite, ensure you observe the following output:

       `Writing data to block 0 at offset 0x0
        Writing data to block 1 at offset 0x20000
        Writing data to block 2 at offset 0x40000`

  6. (Optional) If you have no further use for the Meraki OS, you can
     remove all other UBI volumes on ubi0 (mtd11), including diagnostic1,
     part.old, storage and part.safe. You must not remove the ubi1 ART
     partition (mtd12).

       `for i in diagnostic1 part.old storage part.safe ; do
        ubirmvol /dev/ubi0 -N $i
        done`

  7. Proceed to flash the sysupgrade image via luci, or else download or
     scp the image to /tmp and use the sysupgrade command.

[1] The mentioned images and ubootwrite.py script can be found in this repo:
    https://github.com/clayface/openwrt-cryptid

[2] The modified u-boot sources for the MR42 and MR52 are available:
    https://github.com/clayface/U-boot-MR52-20200629

Signed-off-by: Matthew Hagan <mnhagan88@gmail.com>

zpc0 commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8068-mr42.dts:
```diff
@@ -0,0 +1,179 @@
+
+&soc {
+	mdio_gpio0: mdio {
+		compatible = "virtual,mdio-gpio";
```
Please try to use ipq8064-mdio driver.
92da53b12dce
0d8098548ecb
e42aca06ab13

zpc0 replied:
If `ipq8064-mdio` doesn't works, I think it is ok to use `mdio-gpio`. Thank you for clarification and testing.

clayface replied:
I revisited this issue after discovering that within `mdio0_pins_active`, function had been defined `gpio`. After correcting this to `mdio`, I was able to get this working without the bitbang.

clayface replied:
Returning to this issue, it seems 5.4 is still affected but not 5.10. Two options here; revert back to mdio-gpio, or; stick with the ipq8064-mdio driver and patch back the bit-bang into 5.4, thus no further changes needed when 5.4 support is removed.

zpc0 commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8068-mr42.dts:
```diff
@@ -0,0 +1,179 @@
+
+	aliases {
+		serial1 = &gsbi1_serial;
+		mdio-gpio0 = &mdio_gpio0;
```
If `ipq8064-mdio` works, this can be dropped.

zpc0 commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8068-mr52.dts:
```diff
@@ -0,0 +1,195 @@
+
+	aliases {
+		serial1 = &gsbi1_serial;
+		mdio-gpio0 = &mdio_gpio0;
```
Same as above.

zpc0 commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8068-mr42.dts:
```diff
@@ -0,0 +1,179 @@
+	status = "okay";
+
+	qcom,id = <3>;
+	mdiobus = <&mdio0>;
```
`<&mdio_gpio0>`?

clayface replied:
Either works in this case. Better to use `<&mdio_gpio0>`?

zpc0 replied:
I prefer `mdio_gpio0` because it is the actual mdio bus.

clayface committed:
kernel: add back kmod-leds-tlc591xx

Add back support for the TLC591xx series LEDs which are used in the
ipq806x-based Meraki Cryptid series devices.

This module previously existed for the mvebu platform but was removed
at commit f849c2c83247340d623fdb549e2b75f4c1feea62 due to being enabled
in that platform's kernel config.

Signed-off-by: Matthew Hagan <mnhagan88@gmail.com>

adschm commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8068-cryptid-common.dtsi:
```diff
@@ -0,0 +1,226 @@
+#include "qcom-ipq8064-v2.0.dtsi"
```
Please add an SPDX license identifier.

clayface replied:
I am just about to push some changes, including moving the tlc591xx package/kernel/modules/leds and fixing some whitespace issues. Will add this now then push.

adschm commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8068-cryptid-common.dtsi:
```diff
@@ -0,0 +1,226 @@
+/ {
+	memory {
+		device_type = "memory";
+		linux,usable-memory = <0x41500000 0x1EA00000>;
```
I'd personally prefer lower-case for the hex letters.

adschm commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8068-cryptid-common.dtsi:
```diff
@@ -0,0 +1,226 @@
+	};
+	
+	chosen {
+		bootargs-append = " console=ttyMSM0,115200n8 ubi.mtd=ubi ubi.mtd=art" ;
```
Unneeded space before the semicolon.

adschm commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8068-cryptid-common.dtsi:
```diff
@@ -0,0 +1,226 @@
+
+			partition@0 {
+				label = "sbl1";
+				reg = <0x00000000 0x40000>;
```
Please be consistent with the number of digits. I'd recommend using 7 digits, as that's the size of the biggest number.

Initial offset may be abbreviated to 0x0.

adschm commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8068-mr42.dts:
```diff
@@ -0,0 +1,181 @@
+		pinctrl-0 = <&led_pins>;
+		pinctrl-names = "default";
+
+		led_power_orange: power_orange {
```
If you only have on color per function, you could just use `led_power: power {`
But that's optional, of course.

adschm commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8068-mr42.dts:
```diff
@@ -0,0 +1,181 @@
+	};
+};
+
+
```
Remove second empty line.

adschm commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8068-mr52.dts:
```diff
@@ -0,0 +1,197 @@
+			gpios = <&qcom_pinmux 19 GPIO_ACTIVE_HIGH>;
+		};
+
+		green_lan2 {
```
For the node names, the order is typically inverted compared to the label, so I'd have chosen lan2_green here etc.

adschm reviewed (commented):
It would probably be better to do the kernel changes in separate commits before adding the device(s).

clayface committed:
ipq806x: backport GMAC_AHB_RESET deassert patches

Add backports of the following patches:
"net: stmmac: explicitly deassert GMAC_AHB_RESET" and
"ARM: dts: qcom: add ahb reset to ipq806x-gmac"
Required for Meraki MR42/MR52.

Signed-off-by: Matthew Hagan <mnhagan88@gmail.com>

clayface committed:
ipq806x: add gsbi2_i2c label

gsbi2_i2c is used by the Meraki MR42 so we need to expose a label here.

Signed-off-by: Matthew Hagan <mnhagan88@gmail.com>

clayface commented:
> It would probably be better to do the kernel changes in separate commits before adding the device(s).

Have separated out the patches and fixed the conflict with leds.mk.

zpc0 commented:
Many boards restricts write access to the following partitions.
`SBL1 MIBIB SBL2 SBL3 DDRCONFIG SSD TZ RPM`

Is it possible to add `read-only` to these partitions?

clayface commented:
> Many boards restricts write access to the following partitions.
> `SBL1 MIBIB SBL2 SBL3 DDRCONFIG SSD TZ RPM`
> 
> Is it possible to add `read-only` to these partitions?

Thanks for spotting. Only bootkernel1, bootkernel2 and ubi need RW access.

ziswiler commented on package/kernel/linux/modules/leds.mk:
```diff
@@ -159,3 +159,19 @@ define KernelPackage/leds-uleds/description
+  SUBMENU:=$(LEDS_MENU)
+  TITLE:=LED driver for TLC59108 and TLC59116 controllers
+  DEPENDS:=+kmod-i2c-core +kmod-regmap-i2c
+  KCONFIG:= CONFIG_LEDS_TLC591XX
```
Looks like a spurious space after that colon equal.

clayface replied:
Thanks!

ziswiler commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8068-cryptid-common.dtsi:
```diff
@@ -0,0 +1,236 @@
+
+	uart1_pins: uart1_pins {
+		mux {
+			pins = "gpio51", "gpio52", "gpio53", "gpio54";
```
Are all four UART pins really used?

clayface replied:
There is a hint in the [u-boot source](https://github.com/clayface/U-boot-MR52-20200629/blob/8c6f4c2a84c050555e2dc7036b8287fd9381939d/U-boot.MR52/board/qcom/ipq806x_cdp/ipq806x_board_param.h#L357) that only `gpio51` & `gpio52` may be needed. I have tested and this seems to be all that's needed. Have adjusted the PR accordingly. Thanks!

robimarko replied:
By default on this reference board series 4 pins are used as its supposed to be HSUART, however its possible that only 2 pins for regular UART are used

ziswiler reviewed (approved):
Looks good!

clayface committed:
ipq806x: add GSBI nodes to ipq8064-dtsi-addidions

Rather than having separate patches for each GSBI node added, this patch
consolidates the existing GSBI1 patch into
083-ipq8064-dtsi-additions.patch. In addition, GSBI6 and GSBI7 I2C nodes,
required for the MR42 and MR52 respectively, are added.

Signed-off-by: Matthew Hagan <mnhagan88@gmail.com>

clayface commented:
Discovered the `tlc591xx` tricolour is on `gsbi6`, meaning we can get rid of the i2c bit-bang. Rather than creating another new patch to add the gsbi6 node I have consolidated `gsbi6`, `gsbi7` and the existing `gsbi1` patch within `083-ipq8064-dtsi-additions.patch`. @robimarko, @adschm if this is not preferred please let me know.

zpc0 commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8068-cryptid-common.dtsi:
```diff
@@ -0,0 +1,241 @@
+
+#include "qcom-ipq8064-v2.0.dtsi"
+
+#include <dt-bindings/soc/qcom,tcsr.h>
```
Is this required?
I could not find `TCSR_USB_SELECT_USB3_x` or `IPQ806X_TCSR_REG_x_ADM_MUX_SEL` in dts(i).

clayface replied:
Had mistaken this for `qcom,gsbi.h`, which is already included in ipq8064.dtsi. This has been removed, likewise from the mr42/mr52 dts files. Thanks!

clayface commented:
`ipq806x: explicitly deassert GMAC_AHB_RESET` has been fixed. Previously if multiple gmacs were used, as in the MR52 and several other devices, an oops would occur due to `devm_reset_control_get` requesting exclusive access to the same resource from each gmac. This has been changed to `devm_reset_control_get_shared` which corrects the issue.

keleticsaba commented:
If nothing goes wrong, this weekend I will try to install it on my mr42, and will post review.
Thank You!

clayface commented:
The `GMAC_AHB_RESET` kernel patches have been applied upstream ([1](https://git.kernel.org/pub/scm/linux/kernel/git/qcom/linux.git/commit/?h=dts-for-5.15&id=e67f325e9cd67562b761e884680c0fec03a6f404) and [2](https://git.kernel.org/pub/scm/linux/kernel/git/qcom/linux.git/commit/?h=dts-for-5.15&id=f95c4c56d65225a537a2d88735fde7ec4d37641d)). The relevant commit has been modified to add these as backports to 5.4 and 5.10.

hauke commented:
Thank you for the patches, I applied them to master in 529eac53719ab2b50a6e4585622c1119db5095b4, cef420e8f75206725d94723e0b8a9d14f2a55a20, 771691ec839b8e6925b5cd1d1a34a7ba67c0313f, 6e5b8c6300019e8148008060dfbabb008c2a5ec5 and 67f52012760ce553ba6393aaba77359a458deeda
